packages/playground/src/utils/pricing_calculator.ts

```diff
@@ -71,3 +71,44 @@ export function computePackageColor(packageName?: string): string {
       return "#f3f3f3";
   }
 }
+
+/**
+ * Calculates the additional TFT balance needed for a package upgrade
+ *
+ * This function determines how many additional TFT tokens are needed for a
+ * package upgrade, considering the current price, any applied discount,
+ * and the user's existing balance.
+ *
+ * @param priceTFT - The current price in TFT tokens
+ * @param appliedDiscount - The discount percentage (0-100) applied to the package
+ * @param needUpgrade - Flag indicating if an upgrade is required
+ * @param currentBalance - The user's current TFT balance
+ * @returns The additional TFT balance needed to perform the upgrade, or 0 if no upgrade is needed or balance is sufficient
+ * @throws Error if input parameters are invalid
+ */
+export function calculateUpgradeBalanceNeeded(
+  priceTFT: number,
+  appliedDiscount: number,
+  needUpgrade = false,
+  currentBalance = 0,
+): number {
+  if (!needUpgrade) return 0;
+
+  if (priceTFT < 0) {
```

Comment:
also check if it is a number i.e not a string

Reply:
Is that needed? The argument type is a number already, this should give a compile-time error